### PR TITLE
scheduler: stopped-yet-running allocs are still running

### LINKED ISSUE #10440: Allocations should not overlap in execution
### Nomad version

Nomad v1.1.0-dev (19d58ce4fc756032d719a3a1b2ba1245464f60e2)

### Operating system and Environment details

Linux devbox 5.8.0-23-generic #24~20.04.1-Ubuntu SMP Sat Oct 10 04:57:02 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

### Issue

I have a service deployed on nomad that seems to occasionally run into `bind: address already in use`. I am using a static port so this suggests that there's multiple tasks running concurrently.

My expectation is that `nomad stop` would mean that tasks have fully exited, so that subsequent `nomad run`s don't contend with shared system resources like ports.

The service I'm working with handles graceful shutdown so sometimes there's a bit of a delay until it fully exits. Regardless, it seems that `nomad stop` doesn't block on tasks exiting and replacement tasks can be scheduled right away.

### Reproduction steps

Submit job

```
nomad run job.hcl
```

Confirm functionality

```
curl localhost:6900 # hello world
```

Then restart

```
nomad stop bug && nomad run job.hcl
```

#### Expected Result

I expected the second allocation to start after the first one exits.

#### Actual Result

Both allocations overlap in execution. The second alloc is started immediately despite the first allocation is still shutting down.

Checking the second allocation reveals a conflict:

```
$ nomad logs -stderr ef9696c9
2021/04/23 18:11:05 listening on addr 127.0.0.1:6900
2021/04/23 18:11:05 listen tcp 127.0.0.1:6900: bind: address already in use
```

### Job 

Simple Go http server that handles graceful but slow shutdown:

```go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"
)

func main() {
	var addr = flag.String("addr", "127.0.0.1:8080", "address")
	flag.Parse()

	go func() {
		http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
			fmt.Fprintf(w, "hello world")
		})
		log.Println("listening on addr", *addr)
		log.Fatal(http.ListenAndServe(*addr, nil))
	}()

	c := make(chan os.Signal, 1)
	signal.Notify(c, syscall.SIGINT, syscall.SIGTERM)
	sig := <-c
	log.Println("got signal", sig)
	time.Sleep(10 * time.Second)
	log.Println("exiting now")
	os.Exit(0)
}
```

Build it

```
go build -o /tmp/server server.go
```

Set up job.hcl

```hcl
job "bug" {
  datacenters = ["dc1"]
  type        = "batch"

  group "bug" {
    network {
      port "foo" { static = 6900 }
    }
    task "bug" {
      driver      = "raw_exec"
      kill_signal = "SIGTERM"
      config {
        command = "/tmp/server"
        args = ["-addr", "${NOMAD_IP_foo}:${NOMAD_HOST_PORT_foo}"]
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by notnoop**
_This is a straw-man PR to kick the discussion going. Needs more testing; before merging._

Updates the scheduler so that it treats the alloc as running and using its resources until it truly completes as reported by the client. Typically, an alloc is terminates quickly upon being stopped, but an alloc may take a "long" time to properly shutdown and free up exclusive resources: e.g. ports, CSI/host volumes.

Previously, the scheduler optimistically treats stopped-yet-running allocs as terminated and may schedule an alloc in their place. This introduces a widow where both stopped-yet-running and new allocs are running on the client. During the window, the client resources can be oversubscribed unexpectedly (with OOM trigger) or lead to port/volume collision. Avoiding such window seems like the correct behavior.

I conjecture in a typical cluster, the schedule will simply place the alloc to a new client without user visible downside. For some constrained jobs, the alloc start may delayed until the previous alloc stopped.

### An Alternative

Alternatively, the client may order alloc operations so it starts new allocs only after conflicting stopped allocs complete. Ordering the operations is more complex to implement. It also delays starting the task unnecessarily - the new alloc would have started faster had it be placed on another client without a conflicting alloc.

Fixes #10440

### CODE REVIEW

**Comment by tgross** [comment]
@notnoop how would this interact with [ephemeral disk migrations](https://www.nomadproject.io/docs/job-specification/ephemeral_disk#migrate)? My understanding is that requires we have the alloc with a desired state of stopped on the server but without the alloc runner having stopped on the client.

**Comment by notnoop** [comment]
> how would this interact with ephemeral disk migrations? 

I believe their logic remains untouched, but I'll test.  This PR changes only changes the resource accounting for a node and whether an alloc may be placed on a node or not.

The other issue that should be tested more extensively is pre-emption. The server should discount the resources of alloc that's to be evicted, and the client should sequence it so the evicted alloc completes before the new one starts.

File: nomad/structs/funcs.go [hunk 1]
```diff
@@ -104,7 +104,7 @@ func AllocsFit(node *Node, allocs []*Allocation, netIdx *NetworkIndex, checkDevi
 	// For each alloc, add the resources
 	for _, alloc := range allocs {
 		// Do not consider the resource impact of terminal allocations
-		if alloc.TerminalStatus() {
+		if alloc.ClientTerminalStatus() {
 			continue
 		}
```
Comment (schmichael) [inline nomad/structs/funcs.go:109]: I'm not sure this code is ever hit. The 2 places we call AllocsFit both already filter down to just non-terminal allocs as far as I can tell.

The call in the plan applier calls AllocsByNodeTerminal(false) to generate the list of allocs, so _all_ terminal allocs are already removed AFAICT.

The call in the worker is also passed a list of allocs from AllocsByNodeTerminal(false).

`AllocsByNodeTerminal` itself makes this all pretty tricky to trace because it's difficult to know the provenance of any given `[]*Allocation` and how its been populated and filtered.

I think leaving this filter logic in place is ok as a defensive measure, but I think we need to ensure the half-terminal allocs aren't filtered out "upstream."

**Review by schmichael — CHANGES_REQUESTED** [review]
Great work coming up with a solution for #10440. I think your assessment of the bug and that this is an optimal solution is correct.

## TODO

Probably obvious since this is marked `Draft` but just to be clear it needs at least one unit test to assert the behavior on the scheduler side, and then an e2e test to ensure a well-behaved client interacts as expected with this change.

## Client-coordinated :-1: 

We could fix this on the client, but I don't think there's any benefits there. It would introduce a lot of tricky additional logic to the client and could slow down time-to-running for an allocation to `max(kill_timeout)` of every job in the cluster! That could be *minutes* of startup latency!

So I think client-coordinated collision detection is strictly worse than this server-coordinated approach.

## Server-coordinated (this) risk

The big risk of this approach that occurs to me is that it will exacerbate [`forever pending`](https://github.com/hashicorp/nomad/issues?q=is%3Aissue+is%3Aopen+pending+forever) issues. Currently if an allocation is stuck in `ClientStatus=pending` you can often stop and restart the job to get out of that tricky situation: since the allocation will be `DesiredStatus=stop`, the scheduler will ignore its resource utilization. This would obviously change that and `ClientStatus=pending` allocations would forever take up cluster resources until cleaned out.

I think we should at least consider (RFC?) escape hatches for `ClientStatus=pending` and/or tasks that exceed their `kill_timeout` before merging this PR. I don't want to make those efforts a hard blocker, but I don't want to make "poorly behaved clients" (`forever pending`, tasks that outlive their `kill_timeout`, etc) an even worse issue for operators by merging this without some consideration to remediating these extremely difficult to track down bugs.

**Comment by benbuzbee** [comment]
We have been running this patch in our cluster since it was released and it's solved this problem for us; is there anyway we can help contribute to confidence in merging something?

**Comment by schmichael** [comment]
Thanks for the report @benbuzbee! It's very helpful in cases like this where a change has pretty subtle yet far reaching implications. We'll consider getting this into main.

**Comment by schmichael** [comment]
This is officially a WIP that I intend to get merged. Still needs quite a bit more testing, but upon investigation I think most of my concerns were unfounded.

> The big risk of this approach that occurs to me is that it will exacerbate [forever pending](https://github.com/hashicorp/nomad/issues?q=is%3Aissue+is%3Aopen+pending+forever) issues.The big risk of this approach that occurs to me is that it will exacerbate [forever pending](https://github.com/hashicorp/nomad/issues?q=is%3Aissue+is%3Aopen+pending+forever) issues.

I have observed cases where a `forever pending` replacement alloc causes a permanent `plan for node rejected` for a node, so at least that `forever pending` case would be likely be *fixed* by this code by virtue of not trying to schedule it to begin with.

Many other `forever pending` cases require operator intervention to use either `alloc stop`, `alloc signal`, or in extreme cases draining and wiping the node. I don't think the scope of this PR needs to be enlarged to try to fix all of those cases. As long as it doesn't make them _worse_ than it's worth them for the obvious correctness benefit laid out in #10440.

Next steps are:
- [ ] Audit all uses of `TerminalStatus()` in `scheduler/`
- [ ] More unit tests
- [ ] An e2e test
- [ ] Manually testing `sticky=true` and/or `migrate=true` if e2e tests don't cover them
- [ ] Add changelog
- [ ] Add upgrade guide mention so users are aware of the change
- [ ] Consider an internals/scheduler doc update

**Comment by schmichael** [comment]
tl;dr -> Did some manual testing and confirmed this does not interfere with migrate=true and sticky=true. This patch does not interfere with the existing alloc replacement (due to deployments or preemption) code paths.

If you simply redeploy a job using those, the replacement allocs are created atomically (as part of the same plan) as the old allocs are created. This is the same as the old behavior and optimal as in the case of a deployment the [Nomad client agents handle waiting](https://github.com/hashicorp/nomad/blob/v1.3.2/client/allocwatcher/alloc_watcher.go#L49-L54) for the original to exit before allowing the replacement to start. (Preemption also uses this for "original" and "replacement" allocs of 2 different jobs. The scheduler *must* make the stop-old/place-new placement atomically otherwise another placement could interleave with the stopping-of-old and placing-of-new.)

But that's not what #10440 is about or what this impacts: this is all about when the stop and placement *are* discrete operations. First the job is stopped. Then the job (or a job using the same resources) is started.

In this case the Nomad client agent's waiting code is completely bypassed as there's no PreviousAllocation set. From the server's perspective there's no causality between the old alloc using a resource and the new alloc wanting it...

...so Mahmood's patch closes this gap by making the scheduler calculate resource usage based on _what's actually running on the client_ (based on ClientStatus), not what the server _wants_ to be running on the client (DesiredStatus | ClientStatus).

**Comment by schmichael** [comment]
Some attempts at trying to visually demonstrate the change:

## Original behavior

```mermaid
gantt
    title Non-overlapping allocations
    dateFormat  x
    axisFormat %s
    section Job 1
    Registered       :j1, 0, 2s
    run | pending    :j1pending, after j1, 2s
    run | running    :j1running, after j1pending, 2s
    stop | running   :j1stopping, after j1running, 2s
    stop | stopped   :j1stopped, after j1stopping, 2s

    section Job 2
    Registered       :j2, after j1pending, 1s
    Blocked          :j2blocked, after j2, 1s
    run | pending    :j2pending, after j1running, 1s
    run | running    :j2running, after j2pending, 1s
```

## Implemented behavior

```mermaid
gantt
    title Non-overlapping allocations
    dateFormat  x
    axisFormat %s
    section Job 1
    Registered       :j1, 0, 2s
    run | pending    :j1pending, after j1, 2s
    run | running    :j1running, after j1pending, 2s
    stop | running   :j1stopping, after j1running, 2s
    stop | stopped   :j1stopped, after j1stopping, 2s

    section Job 2
    Registered       :j2, after j1pending, 1s
    Blocked -->      :j2blocked, after j2, 1s
    run | pending    :j2pending, after j1stopping, 1s
    run | running    :j2running, after j2pending, 1s
```

File: testutil/wait.go [hunk 2]
```diff
@@ -14,6 +14,24 @@ import (
...
+		if success {
+			return
+		}
+
+		if retries == 0 {
+			t.Fatalf("timeout: %v", err)
+		}
+	}
+}
+
```
Comment (schmichael) [inline testutil/wait.go:34]: I just got sick of writing the error callback. LMK if I should stop being lazy and switch to WaitForResult. :sweat_smile:

**Comment by schmichael** [comment]
I'll write an e2e test in a followup PR because I want to get this reviewed ASAP. The job_endpoint test I wrote is unfortunately long, but it does properly exercise this behavior. When run on `main` @ a62a6542e it properly fails at:

```
job_endpoint_test.go:190: timeout: expected 1 allocation but found 2:
```

Which indicates that the 2nd job registration placed an allocation before the old allocation had a terminal client status.

**Review by schmichael — APPROVED** [review]
(I have to approve my own changes to clear my old review status.)

**Review by lgfa29 — APPROVED** [review]
LGTM!

I think you also need to make the same change to the `DeviceAccounter`?
https://github.com/hashicorp/nomad/blob/v1.3.5/nomad/structs/devices.go#L63-L66

A couple of other things I looked into:

- CPU cores will be handled by `AllocsFit` like other resources.
- CSI volumes have a completely different tracking mechanism (claims) and are release via a [`Postrun` allocrunner hook](https://github.com/hashicorp/nomad/blob/v1.3.5/client/allocrunner/csi_hook.go#L118-L122).
- Host volume usage doesn't seem to be tracked? Or at least I couldn't find what it happens 😅 

I haven't check quotas (yet), but that would have to be in a separate PR anyway.

File: testutil/wait.go (hunk 2 above)
Comment (tgross) [reply testutil/wait.go:34]: I was tempted to write a similar helper recently, and I've had a few review comments on other folks' PRs of late where not returning an error in the `return false, nil` case ends up making for hard-to-debug test failures. There are ~600 uses of `WaitForResult*` helpers, so at some point I'd like to take a pass through and see how many we could move to this simpler helper... I think it's a lot.

**Review by tgross — APPROVED** [review]
LGTM

**Comment by tgross** [comment]
> CSI volumes have a completely different tracking mechanism (claims) and are release via a [Postrun allocrunner hook](https://github.com/hashicorp/nomad/blob/v1.3.5/client/allocrunner/csi_hook.go#L118-L122).

Interestingly, with this change and by updating the volumewatcher to query allocs and not just volumes (something I've been stewing on), we might be able to simplify the claim unpublish workflow a bit more.

File: nomad/structs/network_test.go [hunk 3]
```diff
@@ -7,6 +7,7 @@ import (
 	"testing"
 
 	"github.com/hashicorp/nomad/ci"
+	"github.com/stretchr/testify/assert"
```
Comment (shoenig) [inline nomad/structs/network_test.go:10]: just JYI the top-level `test` package from `shoenig/test` is analogous to `testify/assert` (marking failure without stopping)

File: testutil/wait.go (hunk 2 above)
Comment (shoenig) [reply testutil/wait.go:34]: I'd also like to see this pattern cleaned up - but this particular helper doesn't prevent the `nil` error issue, and also I think the concept of `TestMultiplier` can be eliminated in favor of context/timeout

Comment (schmichael) [reply testutil/wait.go:34]: What if I added 
```go
if err == nil {
	t.Fatalf("timeout waiting for test function to succeed (you should probably return a helpful error instead of nil!)")
}
```

The file name and line number will take you to the beginning of the test function that returned `false, nil`. Since the stack frame responsible for returning `false, nil` ... returned ... there's really nothing else we can do here.

**Comment by schmichael** [comment]
> LGTM!
> 
> I think you also need to make the same change to the `DeviceAccounter`? https://github.com/hashicorp/nomad/blob/v1.3.5/nomad/structs/devices.go#L63-L66

Great catch! Fixing.

> I haven't check quotas (yet), but that would have to be in a separate PR anyway.

Good call. Looking into that now as well.

**Comment by schmichael** [comment]
> I think you also need to make the same change to the `DeviceAccounter`? https://github.com/hashicorp/nomad/blob/v1.3.5/nomad/structs/devices.go#L63-L66

Fixed in 102af726cbbaa8ac368b03fcf2279f5711381736. Thanks!

> I haven't check quotas (yet), but that would have to be in a separate PR anyway.

Quotas broadly uses `TerminalStatus() => resources are free`, so it will be more permissive than the rest of the scheduler. What quotas report as in-use may be _lower_ than what the scheduler considers as in-use while allocs are shutting down (eg when allocs are in their `shutdown_delay` and `kill_timeout`). This matches existing quotas behavior and doesn't keep this fix from preventing overlapping resource usage that can cause apps to crash, so I'm going to leave quotas alone.

If at some point in the future we find a compelling reason to make Quotas usage more precise, this will serve as useful prior art.
